### PR TITLE
libtest: Don't panic if unable to spawn thread

### LINKED ISSUE #72482: rustc panics when it can't spawn a thread
I wrote up some public and private strcuts, public methods, and traits in a module called structures.rs

In main.rs, I import structures.rs and write tests in its own separate module (within main.rs) using #[cfg(test)] and #[test] above each test method, which tests some public functions and structs from structures.rs

In both files, the fn main() only contains 1 println! statement.

Compiling both main.rs and structures.rs gives me a few warnings but no errors.

I try to run the unit tests using "cargo test" in the "src" folder of my crate but it doesn't act deterministically. Running "cargo test --verbose" errors with the message:

```
thread '<unnamed>' panicked at 'failed to spawn thread: Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }', /rustc/8d69840ab92ea7f4d323420088dd8c9775f180cd/src/libstd/thread/mod.rs:619:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.43.1 (8d69840ab 2020-05-04) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental

note: some of the compiler flags provided by cargo are hidden

thread 'rustc' panicked at 'src/librustc_codegen_ssa/back/write.rs:1754: panic during codegen/LLVM phase', src/librustc/util/bug.rs:37:26

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.43.1 (8d69840ab 2020-05-04) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental

note: some of the compiler flags provided by cargo are hidden

error: could not compile `ExprCalc`.

Caused by:
  process didn't exit successfully: `rustc --crate-name ExprCalc --edition=2018 src/main.rs --error-format=json --json=diagnostic-rendered-ansi --emit=dep-info,link -C debuginfo=2 --test -C metadata=7d54c5c1fcb15878 -C extra-filename=-7d54c5c1fcb15878 --out-dir /home/rchuppala/ExpressionCalculator/ExprCalc/target/debug/deps -C incremental=/home/rchuppala/ExpressionCalculator/ExprCalc/target/debug/incremental -L dependency=/home/rchuppala/ExpressionCalculator/ExprCalc/target/debug/deps` (exit code: 101)
```

Other times, it works successfully and gives me the result of running the tests. I've done a variety of things, including reinstalling Rust and resourcing the path, and received continued undeterministic results (sometimes working, sometimes erring). Any advice?

rustc --version: rustc 1.43.1 (8d69840ab 2020-05-04)
system: linux CentOS

### PR DESCRIPTION
**Pull request by camelid**
Instead just run the test synchronously.

### CODE REVIEW

File: library/test/src/lib.rs [hunk 1]
```diff
@@ -470,7 +470,10 @@ pub fn run_test(
         let supports_threads = !cfg!(target_os = "emscripten") && !cfg!(target_arch = "wasm32");
         if concurrency == Concurrent::Yes && supports_threads {
             let cfg = thread::Builder::new().name(name.as_slice().to_owned());
-            cfg.spawn(runtest).unwrap();
+            if cfg.spawn(runtest).is_err() {
+                // If we can't spawn a new thread, just run the test synchronously.
```
Comment (camelid) [inline library/test/src/lib.rs:474]: Should we print a message letting the user know?

**Comment by camelid** [comment]
I don't have a ton of experience with closures: what's the best way to fix this?

```
error[E0382]: use of moved value: `runtest`
   --> library/test/src/lib.rs:475:17
    |
473 |             if cfg.spawn(runtest).is_err() {
    |                          ------- value moved here
474 |                 // If we can't spawn a new thread, just run the test synchronously.
475 |                 runtest();
    |                 ^^^^^^^ value used here after move
    |
note: closure cannot be invoked more than once because it moves the variable `desc` out of its environment
   --> library/test/src/lib.rs:451:17
    |
451 |                 desc,
    |                 ^^^^
```

**Comment by jyn514** [comment]
@camelid there is not a simple way to fix it; `run_test_in_process` and `spawn_test_subprocess` both consume a [`TestDesc`](https://doc.rust-lang.org/test/struct.TestDesc.html), but you don't know whether they will be called or not before `spawn()` fails. Actually that caught a different error: this should either assert the failure is `WouldBlock` or only call `runtest()` directly if the variant is `WouldBlock`.

The simplest way is probably to get `run_test_in_process` to take `&mut TestDesc` instead, but that might be tricky.

**Comment by camelid** [comment]
> Actually that caught a different error: this should either assert the failure is `WouldBlock` or only call `runtest()` directly if the variant is `WouldBlock`.

Should be fixed now.

File: library/test/src/lib.rs [hunk 2]
```diff
@@ -470,7 +470,17 @@ pub fn run_test(
         let supports_threads = !cfg!(target_os = "emscripten") && !cfg!(target_arch = "wasm32");
         if concurrency == Concurrent::Yes && supports_threads {
             let cfg = thread::Builder::new().name(name.as_slice().to_owned());
-            cfg.spawn(runtest).unwrap();
+            match cfg.spawn(runtest) {
+                Ok(_) => {}
+                Err(err @ io::Error { .. }) if err.kind() == io::ErrorKind::WouldBlock => {
```
Comment (camelid) [inline library/test/src/lib.rs:475]: Not sure if this is the cleanest way of doing this.

File: library/test/src/lib.rs [hunk 3]
```diff
@@ -470,7 +470,17 @@ pub fn run_test(
...
+                    // If spawning a new thread would block, and blocking was
+                    // requested to not occur, just run the test synchronously.
```
Comment (camelid) [inline library/test/src/lib.rs:477]: I was just copying the `WouldBlock` docs. Do we always request no blocking?

Comment (jyn514) [reply library/test/src/lib.rs:477]: 'no blocking' is not relevant here (https://github.com/rust-lang/rust/issues/46345). The man page for `fork` says:
```
       EAGAIN A  system-imposed  limit  on  the number of threads was encountered.
       EAGAIN The  caller  is operating under the SCHED_DEADLINE scheduling policy
              and does not have the reset-on-fork flag set.  See sched(7).
```
I think we can safely assume this is not a real-time process, so the comment should say
```suggestion
                    // If we're at the thread limit, just run all tests in a single thread.
```

File: library/test/src/lib.rs (hunk 2 above)
Comment (jyn514) [inline library/test/src/lib.rs:475]:
```suggestion
            if let Err(err) = cfg.spawn(runtest) {
                if err.kind() == io::ErrorKind::WouldBlock {
```

**Comment by jyn514** [comment]
@camelid can you remove 'may fix' from the commit message? This PR isn't actually fixing the same problem (although it might fix a related one).

**Comment by andersk** [comment]
There’s probably a better way, but one brute-force strategy that works without much thinking would be to encase the closure in an `Arc<Mutex<Option<…>>>`:

```rust
let runtest = Arc::new(Mutex::new(Some(move || match opts.strategy {
    …
})));
…
cfg.spawn({
    let runtest = Arc::clone(&runtest);
    move || runtest.lock().unwrap().take().unwrap()()
})
…
runtest.lock().unwrap().take().unwrap()()
```

**Comment by hyd-dev** [comment]
@camelid Any update? I would like to take over this if you run out of ~~thread~~ time.

> There’s probably a better way, but one brute-force strategy that works without much thinking would be to encase the closure in an `Arc<Mutex<Option<…>>>`

I think [`Arc::get_mut`](https://doc.rust-lang.org/nightly/std/sync/struct.Arc.html#method.get_mut) and [`Mutex::get_mut`](https://doc.rust-lang.org/nightly/std/sync/struct.Mutex.html#method.get_mut) could also be helpful.

**Comment by andersk** [comment]
`Arc::get_mut` and `Mutex::get_mut` require a unique reference and so can't be used from the thread.

**Comment by hyd-dev** [comment]
> `Arc::get_mut` and `Mutex::get_mut` require a unique reference and so can't be used from the thread.

I meant thay can be used after encountering `ErrorKind::WouldBlock`.

**Comment by jyn514** [comment]
>  EDIT: I think I have a working patch now: 97ca47f.

@hyd-dev I think it's fine to open a new PR and mention this one.

**Comment by jyn514** [comment]
err actually this is waiting-on-review, not waiting-on-author, so I don't know if there's any point to a new PR.

**Comment by camelid** [comment]
> err actually this is waiting-on-review, not waiting-on-author, so I don't know if there's any point to a new PR.

This change doesn't work, so a new PR still makes sense 🙃 

The reason it hasn't been reviewed is (1) it doesn't work and (2) withoutboats is taking a break.

**Comment by hyd-dev** [comment]
New PR: #81546

**Pull request closed without merging**